### PR TITLE
network: physical and macvlan nictype MTU support

### PR DESCRIPTION
**Pull request by tomponline**
This exposes the `network_phys_macvlan_mtu` support in LXC for setting a boot time custom MTU on `physical` and `macvlan` nic device types and updates tests to expect the MTU to be changed if that feature is present.

It also modifies LXD to restore the MTU and MAC of physical devices if changed by the container's settings or from within the container.

Depends on https://github.com/lxc/lxc/pull/2985

### CODE REVIEW

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber this is ready for review now.

The original issue you reported is fixed, that of MTU for vlan passthrough not working.

This also fixes several other issues related to MTU restore once a container is shutdown.

There is an outstanding issue that I have been unable to fix, that is the scenario where a new physical device is hotplugged into a container, has its MTU modified and then the container is shutdown.

Right now I can't figure out a way to restore the MTU after the container has stopped, this will need more discussion with @brauner on the best way to achieve this.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber as it stands, this PR utilises the existing (but renamed) OnPostStop() hook to more accurately reflect when the hook is actually run. 

This allows us to restore MTU and MAC for physical passthrough in situations where the devices are hot-unplugged or where LXC knows about them at boot time . However to cover all common scenarios (specifically the scenario where a device is hotplugged and then container is shutdown) I propose adding a new hook to LXD called `OnStop()` which would be run by LXC's `lxc.hook.stop` which would then be able to switch into the container's network namespace (only) and move the network devices back to the host (as well as restoring their original config) before LXC destroy's the namespace and the kernel moves them back (or deletes them in the case of dummy devices).

I propose the `OnStop` hook should iterate through each nic and then attempt to find the device in the container's netns using `volatile.eth1.name` at which point it can run `nsenter --net=${LXC_NET_NS} ip link set ${NAME} netns ${LXD_PID}` to move the container back to LXD's netns and then restore it's original name, MTU, MAC etc from the other volatile keys added in this PR.

Are you happy with this general approach?

**Comment by tomponline** [comment]
@stgraber I've implemented the physical device MTU/MAC restore based on my proposed approach above.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/container_lxc.go [hunk 1]
```diff
@@ -2711,6 +2711,105 @@ func (c *containerLXC) setupPhysicalParent(deviceName string, m types.Device) (s
 	return hostName, nil
 }
 
+// detachInterfaceRename enters the container's network namespace and moves the named interface
+// in ifName back to the network namespace of the running process as the name specified in hostName.
+func (c *containerLXC) detachInterfaceRename(netns string, ifName string, hostName string) error {
+	pid := os.Getpid()
+
+	_, err := shared.RunCommand("nsenter", fmt.Sprintf("--net=%s", netns), "ip", "link", "set", ifName, "name", hostName, "netns", fmt.Sprintf("%d", pid))
```
Comment (stgraber) [inline lxd/container_lxc.go]: I don't like that we entirely rely on the interface name in the container to have been left intact by the user.

Say you have container that's setup with:
 - eth0 => bridged
 - eth1 => physical

As root in the container I could swap their names, causing LXD to move the veth to the host and then the kernel to move the physical interface to the host.

This now effectively gives me a configured network interface on the host which is attached to a bridge that I can have other containers on, if properly configured, I could use that to take over some specific host traffic.

In general LXD should assume a hostile user in the container and use ways of tracking resources which cannot be modified by the user.

I also think that we should ensure we do an IPv4/IPv6 reset on the interface as it's removed from the container, or at least ensure it's brought down as part of this.

**Comment by stgraber** [comment]
Needs rebasing

File: lxd/container_lxc.go (hunk 1 above)
Comment (tomponline) [reply lxd/container_lxc.go]: @stgraber as discussed in IRC, I will modify this to use forknet and to ensure the nic is down before being removed, and also having its IPs removed.

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber this has been re-implemented using the forknet approach you suggested.

The new forknet command removes IP addresses from interface and brings the interface down before moving it back to the container under its original name.

Regarding the check for whether the ifindex has changed inside the container, I believe this would require me to add a new forknet command to inspect the ifindex of the interface inside the container at boot and at hotplug time so that it can be recorded in volatile data. This would likely need either to execute the network up hook for physical devices too, or add a new function for `lxc.hook.start-host` as the current `OnStart()` function is called as `lxc.hook.pre-start` before networking is setup.

Is this something you would want to do, or are the current protections sufficient? (this brings it inline with hot unplugging devices).

**Comment by tomponline** [comment]
@stgraber I did some more research into this, and I think if it is acceptable to add an OnStartHost() hook to be run as `lxc.hook.start-host` then I should be able to use `c.networkState()` (easily extended to gather ifindex) in order to snapshot all boot time interface's ifindexes in a single hit. I can then store in volatile data associated to the device name (using the device's m["name"] property which is generated by LXD and passed into LXC's `name` interface property to match them up to the container side ifindexes). Then at shutdown and interface removal time I can check whether the existing ifindex inside the container has changed.

File: doc/containers.md [hunk 2]
```diff
@@ -86,20 +86,22 @@ user.\*                                 | string    | -                 | n/a
...
+volatile.idmap.current           | string    | -             | The idmap currently in use by the container
+volatile.idmap.next              | string    | -             | The idmap to use next time the container starts
+volatile.last\_state.idmap       | string    | -             | Serialized container uid/gid map
+volatile.last\_state.power       | string    | -             | Container state as of last host shutdown
+volatile.\<name\>.host\_name     | string    | -             | Network device name on the host (for nictype=bridged or nictype=p2p, or nictype=sriov)
+volatile.\<name\>.hwaddr         | string    | -             | Network device MAC address (when no hwaddr property is set on the device itself)
+volatile.\<name\>.name           | string    | -             | Network device name (when no name propery is set on the device itself)
+volatile.\<name\>.last.created   | string    | -             | Whether or not the network device physical device was created ("true" or "false")
+volatile.\<name\>.last.mtu       | string    | -             | Network device original MTU used when moving a physical device into a container
```
Comment (stgraber) [inline doc/containers.md]: `last_state.mtu` for consistency?

File: doc/containers.md [hunk 3]
```diff
@@ -86,20 +86,22 @@ user.\*                                 | string    | -                 | n/a
...
+volatile.\<name\>.last.hwaddr    | string    | -             | Network device original MAC used when moving a physical device into a container
```
Comment (stgraber) [inline doc/containers.md]: `last_state.hwaddr` for consistency?

File: doc/containers.md [hunk 4]
```diff
@@ -86,20 +86,22 @@ user.\*                                 | string    | -                 | n/a
...
+volatile.idmap.base              | integer   | -             | The first id in the container's primary idmap range
+volatile.idmap.current           | string    | -             | The idmap currently in use by the container
+volatile.idmap.next              | string    | -             | The idmap to use next time the container starts
+volatile.last\_state.idmap       | string    | -             | Serialized container uid/gid map
+volatile.last\_state.power       | string    | -             | Container state as of last host shutdown
+volatile.\<name\>.host\_name     | string    | -             | Network device name on the host (for nictype=bridged or nictype=p2p, or nictype=sriov)
+volatile.\<name\>.hwaddr         | string    | -             | Network device MAC address (when no hwaddr property is set on the device itself)
+volatile.\<name\>.name           | string    | -             | Network device name (when no name propery is set on the device itself)
+volatile.\<name\>.last.created   | string    | -             | Whether or not the network device physical device was created ("true" or "false")
```
Comment (stgraber) [inline doc/containers.md]: `last_state.created` for consistency?

**Comment by stgraber** [comment]
@brauner could you take a look at this one too? Just a quick glance through at the bits that look interesting to you is fine.

**Comment by tomponline** [comment]
@stgraber thanks will make those changes.

I was thinking about how we can avoid the use of LXC's hooks calling into LXD. 

This PR introduces a new hook called `OnStopNS` which is utilises LXC's `lxc.hook.stop` hook.

The key reason I had to introduce this new hook was so that devices can be removed from a container and renamed on the host without relying on the OS to do it, in situations where LXC doesn't know about it (i.e when a device is hot plugged after start).

https://github.com/lxc/lxd/pull/5745/files#diff-082fd0ef6645e509accb07d6a535fee6R2795

This is also why I created `detachInterfaceRename()` and extended `forknet` to allow it to enter a container's netns and move the device back to LXD's netns.

https://github.com/lxc/lxd/pull/5745/files#diff-082fd0ef6645e509accb07d6a535fee6R2758

I was thinking that we could avoid both of these steps and move the rest of the restoration code to the existing `OnStop()` hook if we could rely on LXC to detach and rename hot-plugged devices on shutdown.

LXD uses go-lxc's `AttachInterface()` function to hot-plug devices after start.
https://github.com/lxc/lxd/blob/f5bec32c9c9446561ac0a24c75cbda8b04e442d2/lxd/container_lxc.go#L8322

This in turn calls LXC's `do_lxcapi_attach_interface()` https://github.com/lxc/lxc/blob/e9ab177d3f7029db7a59f5097917d3defe84c28e/src/lxc/lxccontainer.c#L4741

Anyway, I was thinking if we could get `do_lxcapi_attach_interface()` to record somewhere any devices that were hot-plugged (just original host name and destination name at a minimal, but could be extended to also read original mtu and mac for restoration by LXC) then we could update LXC to detach hot-plugged devices when the container is shutdown.

This could just be a text file somewhere (each line being something like `origName newName mtu mac` etc), or perhaps it could communicate with the persistent LXC process and update the state internally?

@brauner is this something that could work?

File: lxd/container_lxc.go [hunk 5]
```diff
@@ -2669,6 +2668,201 @@ func (c *containerLXC) startCommon() (string, error) {
...
+	}
+
+	// If MTU value is specified then there is an original MTU that needs restoring.
+	if c.localConfig[mtuKey] != "" {
+		mtuInt, err := strconv.ParseUint(c.localConfig[mtuKey], 10, 32)
+		if err != nil {
+			return fmt.Errorf("Failed to convert mtu for \"%s\" mtu \"%s\": %v", hostName, c.localConfig[mtuKey], err)
+		}
+
+		if err == nil {
```
Comment (brauner) [inline lxd/container_lxc.go]: That sems redundant given that the above returns on error?

File: lxd/main_forknet.go [hunk 6]
```diff
@@ -58,12 +67,18 @@ void forknet() {
 
 	// Call the subcommands
 	if (strcmp(command, "info") == 0) {
+		pid = atoi(cur);
 		forkdonetinfo(pid);
 	}
+
+	if (strcmp(command, "detach") == 0) {
+		forkdonetdetach(cur);
+	}
```
Comment (brauner) [inline lxd/main_forknet.go]: No `{}` around single line if statements. :)

File: lxd/main_forknet.go [hunk 7]
```diff
@@ -107,3 +129,25 @@ func (c *cmdForknet) RunInfo(cmd *cobra.Command, args []string) error {
...
+		return err
+	}
+
+	// Rename the interface, set it down, and move into parent netns.
+	_, err = shared.RunCommand("ip", "link", "set", ifName, "down", "name", hostName, "netns", lxdPID)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("OK")
```
Comment (brauner) [inline lxd/main_forknet.go]: Is that left over from debugging?

File: lxd/main_nsexec.go [hunk 8]
```diff
@@ -99,6 +99,23 @@ int dosetns(int pid, char *nstype) {
 	return 0;
 }
 
+int dosetns_file(char *file, char *nstype) {
+	__do_close_prot_errno int mntns = -EBADF;
```
Comment (brauner) [inline lxd/main_nsexec.go]: That should probably be named `ns_fd` instead of `mntns`.

File: lxd/main_nsexec.go [hunk 9]
```diff
@@ -99,6 +99,23 @@ int dosetns(int pid, char *nstype) {
...
+
+	mntns = open(file, O_RDONLY);
+	if (mntns < 0) {
+		error("error: open mntns");
```
Comment (brauner) [inline lxd/main_nsexec.go]: Error message is specific to mount namespaces while function appears namespace generic.

File: lxd/main_forknet.go (hunk 6 above)
Comment (tomponline) [reply lxd/main_forknet.go]: @brauner doh, thanks!

File: lxd/main_forknet.go (hunk 7 above)
Comment (tomponline) [reply lxd/main_forknet.go]: @brauner I consciously left it in just so when running it manually there was some output to show that everything worked OK (as if there is an error the contents if printed to stderr too). Happy to remove it though as it doesn't service any functional purpose when not debugging.

File: lxd/main_nsexec.go (hunk 8 above)
Comment (tomponline) [reply lxd/main_nsexec.go]: @brauner will apply same change to dosetns function above I based this on.

File: lxd/main_nsexec.go (hunk 9 above)
Comment (tomponline) [reply lxd/main_nsexec.go]: @brauner will apply same change to dosetns function above I based this on.

**Comment by tomponline** [comment]
@brauner @stgraber requested changes made. Thanks

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by brauner** [comment]
> @stgraber thanks will make those changes.
> 
> I was thinking about how we can avoid the use of LXC's hooks calling into LXD.
> 
> This PR introduces a new hook called `OnStopNS` which is utilises LXC's `lxc.hook.stop` hook.
> 
> The key reason I had to introduce this new hook was so that devices can be removed from a container and renamed on the host without relying on the OS to do it, in situations where LXC doesn't know about it (i.e when a device is hot plugged after start).
> 
> https://github.com/lxc/lxd/pull/5745/files#diff-082fd0ef6645e509accb07d6a535fee6R2795
> 
> This is also why I created `detachInterfaceRename()` and extended `forknet` to allow it to enter a container's netns and move the device back to LXD's netns.
> 
> https://github.com/lxc/lxd/pull/5745/files#diff-082fd0ef6645e509accb07d6a535fee6R2758
> 
> I was thinking that we could avoid both of these steps and move the rest of the restoration code to the existing `OnStop()` hook if we could rely on LXC to detach and rename hot-plugged devices on shutdown.
> 
> LXD uses go-lxc's `AttachInterface()` function to hot-plug devices after start.
> https://github.com/lxc/lxd/blob/f5bec32c9c9446561ac0a24c75cbda8b04e442d2/lxd/container_lxc.go#L8322
> 
> This in turn calls LXC's `do_lxcapi_attach_interface()` https://github.com/lxc/lxc/blob/e9ab177d3f7029db7a59f5097917d3defe84c28e/src/lxc/lxccontainer.c#L4741
> 
> Anyway, I was thinking if we could get `do_lxcapi_attach_interface()` to record somewhere any devices that were hot-plugged (just original host name and destination name at a minimal, but could be extended to also read original mtu and mac for restoration by LXC) then we could update LXC to detach hot-plugged devices when the container is shutdown.
> 
> This could just be a text file somewhere (each line being something like `origName newName mtu mac` etc), or perhaps it could communicate with the persistent LXC process and update the state internally?

No text files please. It would likely mean adding new api helpers to liblxc as the current functions are unsuited. I have discussed and suggested something like this with @stgraber already about 1 year ago. Might be worth bringing up again.

**Pull request merged**